### PR TITLE
feat: add C implementation for `math/base/special/logaddexp`

### LINKED ISSUE #1432: [RFC]: Add C implementation for `@stdlib/math/base/special/logaddexp`
### Description

This RFC proposes adding native C implementation for [`@stdlib/math/base/special/logaddexp`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/logaddexp/).

```c
double stdlib_base_logaddexp( const double x, const double y );
```

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by aman-095**
Resolves #1432.

## Description

> What is the purpose of this pull request?

This pull request adds native C implementation for
`@stdlib/math/base/special/logaddexp`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1432 
-   fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/logaddexp/README.md [hunk 1]
```diff
@@ -85,6 +85,99 @@ for ( i = 0; i < x.length; i++ ) {
...
+
+<section class="usage">
+
+### Usage
+
+```c
+#include "stdlib/math/base/special/logaddexp.h"
+```
+
+#### stdlib_base_logaddexp( x )
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/README.md]:
```suggestion
#### stdlib_base_logaddexp( x, y )
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/README.md [hunk 2]
```diff
@@ -85,6 +85,99 @@ for ( i = 0; i < x.length; i++ ) {
...
+    double x;
+    double y;
+    double v;
+    int i;
+    
+    for ( i = 0; i < 100; i++ ) {
+        x = ( ( (double)rand() / (double)RAND_MAX ) * 200.0 ) - 100.0;
+        y = ( ( (double)rand() / (double)RAND_MAX ) * 200.0 ) - 100.0;
+        v = stdlib_base_logaddexp( x, y );
+        printf( "x: %lf, y: %lf, f(x): %lf\n", x, y, v );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/README.md]:
```suggestion
        printf( "x: %lf, y: %lf, logaddexp(x): %lf\n", x, y, v );
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/benchmark/c/native/benchmark.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+/**
+* Benchmark `log1mexp`.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/benchmark/c/native/benchmark.c]:
```suggestion
* Benchmark `logaddexp`.
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/benchmark/c/native/benchmark.c [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	double x;
+	double y;
+    double v;
+	double t;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 200.0*rand_double() ) - 100.0;
+        y = ( 200.0*rand_double() ) - 100.0;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/benchmark/c/native/benchmark.c]:
```suggestion
		x = ( 200.0*rand_double() ) - 100.0;
		y = ( 200.0*rand_double() ) - 100.0;
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/benchmark/c/native/benchmark.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+/**
+* Runs a benchmark.
+*
+* @return elapsed time in seconds
+*/
+double benchmark() {
+	double elapsed;
+	double x;
+	double y;
+    double v;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/benchmark/c/native/benchmark.c]:
```suggestion
	double y;
	double v;
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/examples/c/example.c [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    double x;
+    double y;
+    double v;
+    int i;
+    
+    for ( i = 0; i < 100; i++ ) {
+        x = ( ( (double)rand() / (double)RAND_MAX ) * 200.0 ) - 100.0;
+        y = ( ( (double)rand() / (double)RAND_MAX ) * 200.0 ) - 100.0;
+        v = stdlib_base_logaddexp( x, y );
+        printf( "x: %lf, y: %lf, f(x): %lf\n", x, y, v );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/examples/c/example.c]:
```suggestion
        printf( "x: %lf, y: %lf, logaddexp(x): %lf\n", x, y, v );
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/include/stdlib/math/base/special/logaddexp.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Computes the natural logarithm of `exp(x) + exp(y).
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/include/stdlib/math/base/special/logaddexp.h]:
```suggestion
* Computes the natural logarithm of `exp(x) + exp(y)`.
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/lib/native.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+// MODULES //
+
+var addon = require( './../src/addon.node' );
+
+
+// MAIN //
+
+/**
+* Returns the positive difference between `x` and `y` if `x > y`; otherwise, returns `0`.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/lib/native.js]:
```suggestion
* Computes the natural logarithm of `exp(x) + exp(y)`.
```

File: lib/node_modules/@stdlib/math/base/special/logaddexp/manifest.json [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    "dependencies": [
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/exp",
+        "@stdlib/math/base/special/log1p",
+        "@stdlib/constants/float64/ln-two"
+    ]
+    }
+]
+}
+
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/manifest.json]: Incorrect indentation, please fix this

File: lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+*/
+
+#include "stdlib/math/base/special/logaddexp.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/exp.h"
+#include "stdlib/math/base/special/log1p.h"
+#include "stdlib/constants/float64/ln_two.h"
+
+/**
+* Computes the natural logarithm of `exp(x) + exp(y).
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c]:
```suggestion
* Computes the natural logarithm of `exp(x) + exp(y)`.
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
This is good, nice progress. Requested a few changes, once done this PR will be ready to merge.

File: lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+* @param x    input value
+* @param y    input value
+* @return	  output value
+*
+* @example
+* double out = stdlib_base_logaddexp( 90.0, 90.0 );
+* // returns ~90.6931
+*/
+double stdlib_base_logaddexp( const double x, const double y ) {
+    double ax;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c]: I am not sure if we shall use `double d` instead, as far as I remember `ax` is used for `abs(x)` as per code conventions.

**Review by Pranavchiku — APPROVED** [review]
There is a minor comment I'll let Athan or Phillip address, rest all looks good to me. Thanks @aman-095!

File: lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c (hunk 11 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c]: We should call it `d` and match the JS implementation.

File: lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    if ( x == y ) {
+        return x + STDLIB_CONSTANT_FLOAT64_LN2;
+    }
+    ax = x - y;
+    if ( ax > 0.0 ) {
+        return x + stdlib_base_log1p( stdlib_base_exp( -ax ) );
+    }
+    return y + stdlib_base_log1p( stdlib_base_exp( ax ) );
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/logaddexp/src/main.c]:
```suggestion
double stdlib_base_logaddexp( const double x, const double y ) {
    double d;
    if ( stdlib_base_is_nan( x ) || stdlib_base_is_nan( y ) ) {
        return 0.0/0.0;
    }
    if ( x == y ) {
        return x + STDLIB_CONSTANT_FLOAT64_LN2;
    }
    d = x - y;
    if ( d > 0.0 ) {
        return x + stdlib_base_log1p( stdlib_base_exp( -d ) );
    }
    return y + stdlib_base_log1p( stdlib_base_exp( d ) );
}
```

**Review by Pranavchiku — APPROVED** [review]
Thanks @aman-095 for your work! I'll recommend you to shift towards slightly more difficult packages :)

**Pull request merged**